src/flag_attn/paged.py

```diff
@@ -0,0 +1,415 @@
+
+        # qk: [PADDED_QUERY_GROUP_SIZE, KV_BLOCK_SIZE]
+        if PADDED_QUERY_GROUP_SIZE == 1:
+            qk = tl.sum(q[:, None, :] * k[None, :, :], axis=2)
+        else:
+            qk = tl.dot(q, k.T, out_dtype=tl.float32)
+
+        qk *= attn_scale
+        qk = tl.where(mask_offset < context_len, qk, float("-inf"))
+
+        # m_ij: [PADDED_QUERY_GROUP_SIZE]
+        m_ij = tl.max(qk, axis=1)
+        m_i_new = tl.maximum(m_i, m_ij)
+
+        # p: [PADDED_QUERY_GROUP_SIZE, KV_BLOCK_SIZE]
+        p = tl.exp(qk - m_i_new[:, None])
+        alpha = tl.exp(m_i - m_i_new)
+        acc *= alpha[:, None]
+
+        # v: [KV_BLOCK_SIZE, HEAD_SIZE]
+        v = tl.load(v_cache_ptr + kv_block_offset, mask=kv_mask, other=0.0)
+
+        if PADDED_QUERY_GROUP_SIZE == 1:
+            acc += tl.sum(p.T[:, :, None] * v[:, None, :], axis=0)
+        else:
+            p = p.to(v.dtype)
+            acc += tl.dot(p, v, out_dtype=tl.float32)
+
+        l_i = l_i * alpha + tl.sum(p, axis=1)
+        m_i = m_i_new
+    acc = acc / l_i[:, None]
+
+    if USE_PARTITIONING:
+        part_offset = (
+            (seq_idx * NUM_KV_HEADS + kv_head_idx)
+            * max_num_partitions
+            * QUERY_GROUP_SIZE
+            + part_idx * QUERY_GROUP_SIZE
+            + padding_group_offset
+        )
+        mask = padding_group_offset < QUERY_GROUP_SIZE
+        tl.store(m_i_ptr + part_offset, m_i, mask=mask)
+        tl.store(l_i_ptr + part_offset, l_i, mask=mask)
+
+    out_offset = seq_idx * stride_o0
+    if USE_PARTITIONING:
+        out_offset += kv_head_idx * stride_o1
+    else:
+        out_offset += kv_head_idx * QUERY_GROUP_SIZE * stride_o1
+    out_offset += (
+        part_idx * stride_o2
+        + padding_group_offset[:, None] * stride_o3
+        + head_offset[None, :] * stride_o4
+    )
+
+    group_mask = padding_group_offset[:, None] < QUERY_GROUP_SIZE
+    tl.store(out_ptr + out_offset, acc, mask=group_mask)
+
+
+@triton.jit
+def _paged_attn_v2_reduce_kernel(
+    out_ptr,  # [num_seqs, NUM_KV_HEADS, QUERY_GROUP_SIZE, HEAD_SIZE]
+    m_i_ptr,  # [num_seqs, NUM_KV_HEADS, max_num_partitions, QUERY_GROUP_SIZE]
+    l_i_ptr,  # [num_seqs, NUM_KV_HEADS, max_num_partitions, QUERY_GROUP_SIZE]
+    tmp_out_ptr,  # [num_seqs, NUM_KV_HEADS, max_num_partitions, QUERY_GROUP_SIZE, HEAD_SIZE]
+    context_lens_ptr,  # [num_seqs]
+    max_num_partitions,  # partition stride
+    stride_o0,
+    stride_o1,
+    stride_o2,
+    HEAD_SIZE: tl.constexpr,
+    QUERY_GROUP_SIZE: tl.constexpr,
+    NUM_KV_HEADS: tl.constexpr,
+    PARTITION_SIZE: tl.constexpr,
+    NUM_PARTITIONS: tl.constexpr,
+):
+    seq_idx = tl.program_id(0)
+    kv_head_idx = tl.program_id(1)
+
+    context_len = tl.load(context_lens_ptr + seq_idx)
+
+    num_partitions = tl.cdiv(context_len, PARTITION_SIZE)
+    group_head_offset = (
+        tl.arange(0, QUERY_GROUP_SIZE)[:, None] * HEAD_SIZE
+        + tl.arange(0, HEAD_SIZE)[None, :]
+    )
+    if num_partitions == 1:
```

Comment:
Is this a dummy case(when no partitioning is used, this kernel is not launched)?